cpp/src/arrow/array/validate.cc

```diff
@@ -180,7 +180,9 @@ struct ValidateArrayImpl {
   bool IsBufferValid(int index) { return IsBufferValid(data, index); }
 
   static bool IsBufferValid(const ArrayData& data, int index) {
-    return data.buffers[index] != nullptr && data.buffers[index]->data() != nullptr;
+    return data.buffers[index] != nullptr &&
+           (data.buffers[index]->is_cpu() ? data.buffers[index]->data() != nullptr
+                                          : true);
```

Comment:
This is basically disabling the check on non-CPU buffers. Rather, you should use `data.buffers[index]->address()`.